src/main/java/org/dataone/hashstore/hashfs/HashFileStore.java

```diff
@@ -220,4 +247,244 @@ protected HashAddress put(InputStream object, String pid, String additionalAlgor
+        if (additionalAlgorithm != null) {
+            logHashFileStore.info(
+                    "HashFileStore.writeToTmpFileAndGenerateChecksums - Adding additional algorithm to hex digest map, algorithm: "
+                            + additionalAlgorithm);
+            extraAlgo = MessageDigest.getInstance(additionalAlgorithm);
+        }
+
+        try {
+            byte[] buffer = new byte[8192];
+            int bytesRead;
+            while ((bytesRead = dataStream.read(buffer)) != -1) {
+                os.write(buffer, 0, bytesRead);
+                md5.update(buffer, 0, bytesRead);
+                sha1.update(buffer, 0, bytesRead);
+                sha256.update(buffer, 0, bytesRead);
+                sha384.update(buffer, 0, bytesRead);
+                sha512.update(buffer, 0, bytesRead);
+                if (additionalAlgorithm != null) {
+                    extraAlgo.update(buffer, 0, bytesRead);
+                }
+            }
+        } catch (IOException ioe) {
+            logHashFileStore.error(
+                    "HashFileStore.writeToTmpFileAndGenerateChecksums - IOException encountered (os.flush/close or write related): "
+                            + ioe.getMessage());
+            throw ioe;
+        } finally {
+            os.flush();
+            os.close();
+        }
+
+        logHashFileStore.info("HashFileStore.writeToTmpFileAndGenerateChecksums - Object has been written to tmpFile");
+        String md5Digest = DatatypeConverter.printHexBinary(md5.digest()).toLowerCase();
+        String sha1Digest = DatatypeConverter.printHexBinary(sha1.digest()).toLowerCase();
+        String sha256Digest = DatatypeConverter.printHexBinary(sha256.digest()).toLowerCase();
+        String sha384Digest = DatatypeConverter.printHexBinary(sha384.digest()).toLowerCase();
+        String sha512Digest = DatatypeConverter.printHexBinary(sha512.digest()).toLowerCase();
+        hexDigests.put("MD5", md5Digest);
+        hexDigests.put("SHA-1", sha1Digest);
+        hexDigests.put("SHA-256", sha256Digest);
+        hexDigests.put("SHA-384", sha384Digest);
+        hexDigests.put("SHA-512", sha512Digest);
+        if (additionalAlgorithm != null) {
+            String extraDigest = DatatypeConverter.printHexBinary(extraAlgo.digest()).toLowerCase();
+            hexDigests.put(additionalAlgorithm, extraDigest);
+        }
+
+        return hexDigests;
+    }
+
+    /**
+     * Moves an object from one location to another if the object does not exist
+     * 
+     * @param source file to move
+     * @param target where to move the file
+     * 
+     * @return fileMoved (boolean) to confirm file is not a duplicate and has been
+     *         moved
+     * @throws IOException       Unable to create parent directory
+     * @throws SecurityException Insufficient permissions to move file
+     */
+    protected boolean move(File source, File target) throws IOException, SecurityException {
+        if (target.exists()) {
```

Comment:
I think we should throw an exception if the file exists.

Reply:
Will refactor based on our discussion above RE: isNotDuplicate